Add and improve docblock typing for static analysis

mdeboer opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| License       | MIT

I noticed that both PhpStorm and Psalm were both not inferring the correct return type for `PassportInterface::getBadge()`, instead they would type it as `BadgeInterface|null`. Essentially this is correct but if a not-null value is returned, it's always an instance of the type of `$badgeFqcn`.

In PhpStorm I at first fixed this with the following snippet in `.phpstorm.meta.php` but Psalm does not pick it up for some reason:

```php
<?php

namespace PHPSTORM_META {
    use Symfony\Component\Security\Http\Authenticator\Passport\PassportInterface;

    override(PassportInterface::getBadge(), type(0));
}
```

So with this commit I propose to add the following docblock to `PassportInterface::getBadge()`:

```php
/**
 * @template T of BadgeInterface
 *
 * @param class-string<T> $badgeFqcn
 *
 * @return T|null
 */
```

This works with both PhpStorm and Psalm but if this causes issues for other IDEs, they can be prefixed with `psalm-`. Not sure what the policy is for docblocks in this case 😄

Commit:
Add return type for getBadge in PassportInterface

fabpot commented:
Sorry, but I don't think we want to merge this change. At least, not without having a discussion about using this new notation in Symfony (AFAIK, we have never used it anywhere else).

nicolas-grekas reviewed (dismissed):
I think this is worth the DX improvement this will provide thanks to autocompletion. Ppl keep adding local inline `@var` annotations in similar situations right now AFAIK. This should help save writing them.

wouterj commented:
I agree with @fabpot that this is something we'll have to discuss (and I'm 100% in favor of this btw). The Psalm team is maintaining https://github.com/psalm/psalm-plugin-symfony/tree/master/src/Stubs to add stub-files with these annotations for precisely this usage (in e.g. property accessor). I think it's much better to move all of these into the Symfony code, but I'm against having some of this in the code and some of this in the stubs.

derrabus commented:
Generics in PHPDoc have broad tooling support. PhpStorm, PHPStan and Psalm support this notation and it improves the DX a lot if we used it. As @wouterj has pointed out, people already build packages to deliver type information that we don't provide. I think it's better if we maintain the type information directly in our PHPDoc blocks.

chalasr commented:
I'm ok to do this as it is modern phpdoc. Hope we can remove it in a future PHP version though.

I think it's worth mentioning that this makes the contract more precise, which should be covered by our BC promise:
This method does not just guarantee to always return a BadgeInterface instance anymore, but an instance of the concrete implementation that is requested via method argument.

fabpot commented:
I'm ok-ish only if this is something "standard" that is understood by many tools (for some definition of many).

mdeboer commented:
> I'm ok-ish only if this is something "standard" that is understood by many tools (for some definition of many).

Like @derrabus mentioned, this has broad support in tooling and is not just for PhpStorm and Psalm.

I am glad to see so many reactions and to be honest, I completely forgot about the stubs repository @wouterj mentioned. I agree with his opinion in that having code in both a third-party stubs repository and in the Symfony code is not ideal.

As this piece of code is marked experimental, I suppose we could make the first steps of merging these annotations from the stubs into the Symfony code where it makes sense.

Thanks all for taking the time! 👍🏻

nicolas-grekas commented:
I personally would be OK to import all annotations that exist on the stubs.
Of course, we should import only the annotations that are supported by many tools (eg not psalm-specific annotations.)

@mdeboer would you mind updating this PR to add them all?

dunglas commented:
I was in the process of opening a similar PR for forms when @nicolas-grekas commented: https://github.com/symfony/symfony/compare/5.x...dunglas:feat/form-psalm-template.

Having such annotations will allow to not introduce "useless" code just to have autocompletion working properly such as proposed in https://github.com/symfony/symfony/pull/40799#issuecomment-819008314.

+1 on my side to have these annotations directly in the code base. It will increase the DX of Symfony for both PHPStorm and Psalm/PHPStan users.

mdeboer commented:
I will update the PR to add all the relevant stubs that are not just for psalm (e.g. without psalm prefix or supported by others with psalm prefix).

Commit:
Add annotations for Forms

Annotations are taken from psalm/psalm-plugin-symfony. Only non-psalm specific annotations are added.

Commit:
Possible fix psalm error on view variables.

Commit:
Annotate IteratorAggregate types

Commit:
Annotate serializer types

Commit:
Annotate Cookie sameSite parameters

Commit:
Fix annotations for CacheInterface

mdeboer commented on src/Symfony/Component/Form/Extension/Core/Type/ChoiceType.php:
```diff
@@ -287,7 +287,7 @@ public function buildView(FormView $view, FormInterface $form, array $options)
         }
 
-        if ($options['multiple'] && !$options['expanded']) {
+        if ($options['multiple'] && !$options['expanded'] && isset($view->vars['full_name'])) {
```
I was unsure about adding this but it can easily be reverted. It seems to me it can't hurt to add this check and as PHPUnit tests seem to pass 🤷🏻 Anyway, I replicated the psalm setup from the Github action locally using Docker and psalm came up with this. Not sure whether it is something my baseline missed or because of changes I made.

nicolas-grekas replied:
I think it's not needed. If I'm wrong, this should be patched in another PR anyway. So yes, please revert.

mdeboer commented on src/Symfony/Contracts/Cache/CacheInterface.php:
```diff
@@ -29,14 +29,16 @@ interface CacheInterface
+     *                                                                        The default (or providing null) is implementation dependent but should
+     *                                                                        typically be 1.0, which should provide optimal stampede protection.
+     *                                                                        See https://en.wikipedia.org/wiki/Cache_stampede#Probabilistic_early_expiration
+     * @param array|null                                           &$metadata The metadata of the cached item {@see ItemInterface::getMetadata()}
```
This was annotated as array but the default for this parameter was `null` making the actual type `array|null`, I fixed it here. Doubt this was intentional, but if it was I'm happy to revert this.

Commit:
Style fixes

Commit:
Generics annotation for DataTransformerInterface

mdeboer commented:
Okay so I added a lot of annotations, many of which had to be adapted to be accepted by more than just Psalm. I'm particularly curious about the Form component docblocks as I never touched that component. I felt like things could be improved but not without further changes and being stuck with Psalm.

Also for some reason the Psalm check now fails but it seems to be a bug in Psalm? When I run it locally using Docker (PHP 8) and replicate the workflow setup, all is fine.

Anyway, please let me know if there's something I can do. Feel free to edit as you please.

mdeboer commented on src/Symfony/Component/Console/Helper/HelperSet.php:
```diff
@@ -89,7 +91,7 @@ public function getCommand()
 
     /**
-     * @return Helper[]
+     * @return \Traversable<Helper>
```
I am not sure whether we should use `\Traversable<Helper>` or `\Traversable<Helper>|Helper[]`. Technically the first is correct but I believe that the latter is a common workaround for IDEs not being able to make sense of the generic `\Traversable`. Though I'm not sure this applies to the `getIterator` method in case the method is not called but the object is iterated over in a `foreach`.

nicolas-grekas commented on src/Symfony/Component/Form/FormFactoryInterface.php:
```diff
@@ -23,7 +23,8 @@ interface FormFactoryInterface
      * @see createBuilder()
      *
-     * @param mixed $data The initial data
+     * @param class-string<FormTypeInterface>|'Symfony\Component\Form\Extension\Core\Type\FormType' $type
```
is the part about `FormType` really needed? what does it provide? can we remove it?

stof replied:
this change looks wrong to me. `class-string<FormTypeInterface>` really means `FormTypeInterface::class`, not "class name of a class extending FormTypeInterface"

mdeboer replied:
I took this from the psalm stubs and what @stof says makes sense. I had to make it a union for it to accept the default value.

But yeh I doubt it adds much for the DX and is a headache to get right. I will remove this.

nicolas-grekas commented on src/Symfony/Component/Form/FormInterface.php:
```diff
@@ -17,6 +17,10 @@
+ *
+ * @template T
+ *
+ * @method void clearErrors(bool $deep)
```
why this?
this should be documented in changelog+upgrade if we really want to add this method to the interface

mdeboer replied:
I have absolutely no clue. This was in the psalm stubs and seemed odd to me, but then again I never touched the Forms component. I will remove it.

nicolas-grekas commented on src/Symfony/Component/HttpFoundation/Cookie.php:
```diff
@@ -71,6 +71,19 @@ public static function fromString(string $cookie, bool $decode = false)
+     * @param bool|null                     $secure   Whether the client should send back the cookie only over HTTPS or null to auto-enable this when the request is already using HTTPS
+     * @param bool                          $httpOnly Whether the cookie will be made accessible only through the HTTP protocol
+     * @param bool                          $raw      Whether the cookie value should be sent with no url encoding
+     * @param self::SAMESITE_*|''|null      $sameSite Whether the cookie will be available for cross-site requests
```
we try to remove docblock when they don't add much value, and we do partial `@param` when only some arguments benefit from the docblock. Please remove the lines that are obvious here.

nicolas-grekas replied:
now I see that `__construct` has the same docblock, so maybe we considered that there is enough value in keep it :)

stof replied:
I think the CS fixer kept it due to having a description.

nicolas-grekas commented on src/Symfony/Component/HttpFoundation/Cookie.php:
```diff
@@ -71,6 +71,19 @@ public static function fromString(string $cookie, bool $decode = false)
+     * @param bool                          $raw      Whether the cookie value should be sent with no url encoding
+     * @param self::SAMESITE_*|''|null      $sameSite Whether the cookie will be available for cross-site requests
+     *
+     * @throws \InvalidArgumentException
```
this should be removed: `@throw` is only for exception that should be caught. This one should not as it highlights an error from the dev.

nicolas-grekas replied:
(let's also remove it from the constructor)

nicolas-grekas commented:
Before merging this PR, we should make `DebugClassLoader` know about the new annotations.
This would fix failures like: `TypeError: Symfony\Component\Form\Tests\Fixtures\FixedDataTransformer::transform(): Return value must be of type ?Symfony\Component\Form\R, string returned`

/cc @fancyweb in case you want to have a look :angel:

stof commented on src/Symfony/Component/DomCrawler/Crawler.php:
```diff
@@ -18,6 +18,8 @@
  *
  * @author Fabien Potencier <fabien@symfony.com>
+ *
+ * @implements \IteratorAggregate<\DomNode>
```
the canonical case of the class name is `DOMNode`

stof commented on src/Symfony/Component/Form/CallbackTransformer.php:
```diff
@@ -11,6 +11,9 @@
 namespace Symfony\Component\Form;
 
+/**
+ * @implements DataTransformerInterface<mixed, mixed>
```
this could be actually a generic class, by providing actual signatures for the callables

stof commented on src/Symfony/Component/Form/Extension/Core/DataTransformer/ChoicesToValuesTransformer.php:
```diff
@@ -17,6 +17,8 @@
 /**
  * @author Bernhard Schussek <bschussek@gmail.com>
+ *
+ * @implements DataTransformerInterface<array, array>
```
should be `string[]` for the values to be more precise

stof commented on src/Symfony/Component/Form/Extension/Core/DataTransformer/DateTimeImmutableToDateTimeTransformer.php:
```diff
@@ -18,6 +18,8 @@
  *
  * @author Valentin Udaltsov <udaltsov.valentin@gmail.com>
+ *
+ * @implements DataTransformerInterface<\DateTimeImmutable|null, \DateTime|null>
```
no need of `|null` here. The interface already turns those types into nullable ones

stof commented on src/Symfony/Component/Form/Extension/Core/DataTransformer/StringToFloatTransformer.php:
```diff
@@ -14,6 +14,9 @@
 use Symfony\Component\Form\Exception\TransformationFailedException;
 
+/**
+ * @implements DataTransformerInterface<string|null, int|float|null>
```
same here about not needing `|null`

stof commented on src/Symfony/Component/Form/Extension/Core/DataTransformer/ValueToDuplicatesTransformer.php:
```diff
@@ -16,6 +16,8 @@
 /**
  * @author Bernhard Schussek <bschussek@gmail.com>
+ *
+ * @implements DataTransformerInterface<mixed, mixed>
```
the transformed value is an `array<string, mixed>` (or even better, make this a generic, implementing `DataTransformerInterface<T, array<string, T>>`)

stof commented on src/Symfony/Component/Form/Extension/Core/Type/FormType.php:
```diff
@@ -26,6 +26,9 @@
 use Symfony\Component\PropertyAccess\PropertyAccessorInterface;
 
+/**
+ * @extends BaseType<self>
```
this change is wrong. The templated type for `BaseType` is the one from `FormTypeInterface`, which is the type of the data of the form. That should be `mixed` here, not `self` (the data will never be a `FormType` instance)

stof commented on src/Symfony/Component/Form/FormBuilderInterface.php:
```diff
@@ -32,8 +35,8 @@ public function add($child, string $type = null, array $options = []);
-     * @param string      $name The name of the form or the name of the property
-     * @param string|null $type The type of the form or null if name is a property
+     * @param string                               $name The name of the form or the name of the property
+     * @param class-string<FormTypeInterface>|null $type The type of the form or null if name is a property
```
this is wrong. `class-string<FormTypeInterface>` does not mean "the name of a class extending FormTypeInterface`. It means "the name of FormTypeInterface"

wouterj replied:
I think this is what does cover the correct meaning:

```
@template T of FormTypeInterface

@param class-string<T>|null $type
```

stof commented on src/Symfony/Component/Form/FormTypeExtensionInterface.php:
```diff
@@ -53,7 +63,7 @@ public function configureOptions(OptionsResolver $resolver);
      * Gets the extended types.
      *
-     * @return string[]
+     * @return iterable<class-string<FormTypeInterface>>
```
this is wrong

stof commented on src/Symfony/Component/Form/FormView.php:
```diff
@@ -15,11 +15,17 @@
     /**
      * The variables assigned to this view.
+     *
+     * @var array{value: ?T, attr: array<array-key, mixed>}&array<string, mixed>
```
As reported in the psalm plugin repo, this is wrong. The `value` variable is not guaranteed to be of type `T` (that depends on the data transformers being applied).

stof commented on src/Symfony/Component/Form/ReversedTransformer.php:
```diff
@@ -18,6 +18,8 @@
  *
  * @author Bernhard Schussek <bschussek@gmail.com>
+ *
+ * @implements DataTransformerInterface<mixed, mixed>
```
This could actually be more precise, by making it a generic class implementing `DataTransformerInterface<T, R>` and expecting a `DataTransformerInterface<R, T>` in the constructor.

stof commented on src/Symfony/Component/Serializer/Normalizer/DenormalizerInterface.php:
```diff
@@ -27,12 +27,14 @@ interface DenormalizerInterface
+     * @param string|null                  $format  Format the given data was extracted from
+     * @param array                        $context Options available to the denormalizer
+     *
+     * @return TObject|mixed
```
from a type point of view, that's the same than `mixed`, so this does not bring any value

mdeboer commented:
Hey guys quick update, really busy week but I read all your comments, thank you all! I will see to it this weekend 👍🏻

Commit:
Revert "Annotate serializer types"

This reverts commit a5517de8

Commit:
Form related annotation fixes

Commit:
Fix DOMNode annotation

Commit:
Remove unneeded InvalidArgumentException annotation from Cookie

mdeboer commented:
So I made some fixes, hopefully this ticks all the boxes. Unfortunately fabbot complains about a comment about the return value in `Symfony\Component\Form\Extension\Core\DataTransformer::transform`, wanting to replace `/** @var string[] */` with for example `/* @var string[] */`. Would it help to make it a multi-line comment?

I haven't been able to set up PhpStorm for this project properly yet I reckon but I really haven't got much time at the moment to deep dive into getting all this set up properly.

stof commented on src/Symfony/Component/Form/Extension/Core/DataTransformer/DateTimeZoneToStringTransformer.php:
```diff
@@ -42,6 +44,7 @@ public function transform($dateTimeZone)
                 throw new TransformationFailedException('Expected an array of \DateTimeZone objects.');
             }
 
+            /** @var string[] */
```
the reason why fabbot does not like this is that this is not a valid inline phpdoc variable declaration: it does not have the variable name (which is kind of expected, as this is not applied on a variable declaration). This should be removed IMO

mdeboer replied:
It's a bit of a slippery slope about how far to go to help type hint things and please static analysers. This was one of those things.

I'll remove it 👍🏻

OskarStark commented:
Are you able to finish this PR? Thanks

zmitic commented:
@mdeboer There is another file for PostSubmit event: https://github.com/psalm/psalm-plugin-symfony/blob/master/src/Stubs/common/forms/PostSubmitEvent.stubphp

mdeboer commented:
> Are you able to finish this PR? Thanks

Sorry, really busy weeks. I might give this another go at the end of this week as things probably slow down.

> @mdeboer There is another file for PostSubmit event: https://github.com/psalm/psalm-plugin-symfony/blob/master/src/Stubs/common/forms/PostSubmitEvent.stubphp

Okay, probably missed it. I will see if that was intentional or not and add it accordingly 👍🏻

stof commented:
I don't think that those generic annotations should be added in the form component. Most of my review comments about types being wrong relate to the fact that the Form component is very flexible, which makes it very hard to define type definitions for that (adding a data transformer changes the view data or the model data for that form type for instance)

zmitic commented:
@stof 

>I don't think that those generic annotations should be added in the form component.

If unused, they would still work as before (mixed). But adding them prevents lots of errors and ``@var`` annotations.

The only (somewhat) drawback is the need for this:

```php
$form = $this->createForm(UserType::class);   //   must use 'empty_data' callable to populate

// submit, handle...
if ($user = $form->getData()) {    // must be this way because Form::getData() returns nullable T

    // here psalm **knows** it is instance of User
    $this->dispatch(
        new SomeUserMessage($user);
    );
}
```
Message:

```php
class SomeUserMessage
{
    private string $id;

    public function __construct(User $user)
    {
        $this->id = $user->getId();
    }
}
```

nicolas-grekas commented:
@mdeboer the code base evolved a lot :grimacing: Can you please rebase?

stof commented:
@zmitic the issue is that you are not adding generic types only for the `Form` instance, but for many parts of the components, and those **don't match the behavior of the component**.

A Form instance has **3** types associated to it (the model data, the normalized data and the view data).
The various data transformers are about converting between those, but they are not `DataTransformerInterface<TModel, TNormalized>`. They are a *chain* of transformers being composed, when the input of one of them is the output of the previous one.
Properly defining generic types for the form building API is probably impossible, as the FormBuilder API is about mutating an existing object, and so adding transformers is changing the data types of that instance (which is something that type systems cannot represent AFAIK), and event listeners added on it are expected to use the final data types anyway (even if they are configured before the transformers).

Having wrong generic types will be worse for static analyzers than not having them at all.

fabpot commented:
/cc @derrabus

mdeboer commented:
Thing is I have very, very limited time and in regard of the Form stuff, I never ever touched it so I really have no clue with what is the correct way to proceed there.

This only really started as I noticed that the PassportInterface was lacking docblocks with generics. Though I'd love to be able to help, I simply cannot afford to put work aside and focus on this, specially with the holidays and the on-going crisis. Trying hard to not end up in a burn-out.

If anyone would like repo access, please let me know and I'll grant you access. Also feel free to submit pull requests, I'd be happy to merge them so they end up in this PR.

Take care all and happy holidays!

derrabus commented:
@mdeboer No worries. I intend to take over this PR for 6.1 because I am probably the one who caused most of the conflicts. 🙂

fabpot commented:
@derrabus Friendly ping here :)

wouterj commented:
Hi @mdeboer! I tried to reignite this pull request, but unfortunately there are just too many conflicts with version 6.2 (given we've added lots of native types and PHPdoc for iterator generics in 5.4/6.0, most of the changes in this PR conflict).

So I've decided that it is best to start all over in a new PR: https://github.com/symfony/symfony/pull/47016 I've kept your original commit (adding a template to passport) and the discussions in this PR were very useful in order to judge which changes are "quick wins" and which need a more thorough review (the new PR only contains the quick wins).

I hope you can understand and hopefully we'll see more contributions from you in the future. We've now learned the lesson (once again) that big PRs are very hard to merge... :(

mdeboer commented:
No worries @wouterj I completely understand! I am glad it was of use 😁 I appreciate your follow up.
